### ISSUE TITLE
SQL Internal Error adding image to pool

### ISSUE THREAD

**Issue by diamondsw** [shish/shimmie2#961]
**Server Software**
More to the point, I'm running this in Docker using the "shish2k/shimmie2:latest" image, and "postgres:11-alpine" for the database. I assume something in the update 3 days ago (2023-08-20) has a regression.

about:
  title: "HomeBooru"
  theme: "danbooru2"
  url: "http://booru.redacted.com/index.php?q="

versions:
  shimmie: "2.10.0-alpha"
  schema: 21
  php: "8.2.7"
  db: "pgsql PostgreSQL 11.21 on x86_64-pc-linux-musl, compiled by gcc (Alpine 12.2.1_git20220924-r10) 12.2.1 20220924, 64-bit"
  os: "Linux 989667cbfe21 5.10.0-21-amd64 #1 SMP Debian 5.10.162-1 (2023-01-21) x86_64"
  server: "PHP 8.2.7 Development Server"

extensions:
  core: ["admin","alias_editor","bbcode","comment","download","et","ext_manager","four_oh_four","handle_pixel","help_pages","image","index","media","mime","setup","static_files","system","tag_edit","tag_list","upgrade","upload","user","user_config","view"]
  extra: ["auto_tagger","bulk_add_csv","favorites","home","pools","post_titles","random_image","random_list","rating","regen_thumb","relationships","tag_categories"]
  handled_mimes: ["image/jpeg","image/gif","image/png","image/webp"]

stats:
  images: 19401
  comments: 0
  users: 2

media:
  memory_limit: "8.0MB"
  disk_use: "182.7GB"
  disk_total: "196.7GB"

thumbnails:
  engine: "convert"
  quality: 75
  width: 192
  height: 192
  scaling: 100
  mime: "image/jpeg"

**What steps trigger this bug**
1. Go to 'Pools'
2. Click on an existing pool
3. Search for something to add to the pool.
4. From the result list, check at least one of the items and click "Add Selected".
5. Acknowledge the "Are you sure..." prompt. (I'd really like to disable this/add a "never ask again" check)

**What did you expect to happen?**
The image should have been added to the pool and the page redirected to the pool listing.

**What actually happened?**

The action returned with a page titled `Internal Error - SCore-2.10.0-alpha` and the following content:
```
Internal Error

Message: SQLSTATE[42703]: Undefined column: 7 ERROR: column "lastupdated" of relation "pools" does not exist LINE 2: ...ELECT COUNT(*) FROM pool_images WHERE pool_id=$1),lastupdate... ^

Query:

Version: 2.10.0-alpha (on 8.2.7)

Stack Trace:

#0 /app/core/database.php(182): Shimmie2\Database->_execute()
#1 /app/ext/pools/main.php(980): Shimmie2\Database->execute()
#2 /app/ext/pools/main.php(968): Shimmie2\Pools->update_count()
#3 /app/ext/pools/main.php(718): Shimmie2\Pools->add_post()
#4 /app/core/send_event.php(153): Shimmie2\Pools->onPoolAddPosts()
#5 /app/ext/pools/main.php(367): Shimmie2\send_event()
#6 /app/core/send_event.php(153): Shimmie2\Pools->onPageRequest()
#7 /app/index.php(87): Shimmie2\send_event()
#8 {main}
```

**Comment by shish** [comment]
Argh, yeah - the pools extension had some messed up database update paths, so depending on exactly when it was installed and which database was used, the table would have different columns with different settings D: After #942 it should be fixed for new installs, but needs some manual fixes to bring existing installs into consistency >.<

For a postgres-based install which is missing the lastupdated column, I *think* this SQL should get it back in sync: `ALTER TABLE pools ADD COLUMN lastupdated TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP`

**Comment by diamondsw** [comment]
I probably botched something, as for the moment there's no change:

```
me@docker shimmie $ docker exec -it shimmie-sql psql -U shimmie -W shimmie
Password: 
psql (11.21)
Type "help" for help.

shimmie=# ALTER TABLE pools ADD COLUMN lastupdated TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP
shimmie-# \q
me@docker shimmie $ docker-compose down && docker-compose up -d
Stopping shimmie     ... done
Stopping shimmie-sql ... done
Removing shimmie     ... done
Removing shimmie-sql ... done
Removing network shimmie_default
Creating network "shimmie_default" with the default driver
Creating shimmie-sql ... done
Creating shimmie     ... done
me@docker shimmie $ 
```

But no change when adding the image to the pool.

Once again, I feel you - it sucks when you're a fair ways into a project and you need to extend some core data structure (database or otherwise) and you juts know it's going to be a support nightmare from then on.

The way I usually handle it is to add a migration check when my app launches - if it finds an older setting, table, etc that should be changed, it updates it and moves on with the new code. That way I never worry about having old data structures hanging around for this reason. Is that possible for you, given the complexity of the codebase and the installs you have to support?

**Comment by shish** [comment]
Restarting the server shouldn't be needed

I think the change never happened because you need to add a semicolon to the end of the line - the `shimmie-# ` (as opposed to `shimmie=#` is postgres' "please type more" prompt)


> if it finds an older setting, table, etc that should be changed, it updates it and moves on with the new code

That is done indirectly - each extension has a version number, and when changing the database schema, the idea is to bump the version number and add a small bit of SQL to migrate from the previous version to the current one. Unfortunately the pools extension wasn't doing this properly, and so "install the current version" and "install version 1 and then upgrade to the current version" ended up having different results. For bonus fun, some of the versions were using mysql-specific features, but you'd only hit those if you installed a specific version then upgraded, not if you did a clean install...

A more robust approach would be to directly look at the columns (eg instead of asking "is this extension on version 3?", ask "does this table have a lastupdated column?") - but last time I checked (a few years ago, maybe things are better now) not every database supported that :( If there is reliable support for viewing table schemas for all the database engines, then I'd gladly swap over to doing things that way 🤔

**Comment by diamondsw** [comment]
> Restarting the server shouldn't be needed

I figured as much; it was a hail mary when I saw the initial line hadn't worked...

> I think the change never happened because you need to add a semicolon to the end of the line - the `shimmie-# ` (as opposed to `shimmie=#` is postgres' "please type more" prompt)

...and that was why. Yep, one extra semicolon and it worked great.

> not every database supported that :( If there is reliable support for viewing table schemas for all the database engines, then I'd gladly swap over to doing things that way 🤔

I have been kicking myself for decades for not taking databases when I was in CS, although fragmentation like this has always made me a bit leery of diving in. Best of luck in managing such insanity. :)

**Comment by shish** [comment]
> I have been kicking myself for decades for not taking databases when I was in CS, although fragmentation like this has always made me a bit leery of diving in. Best of luck in managing such insanity. :)

Yeah, SQL looks like such a nice standard at first glance... but then you learn that every database engine has _completely_ different ways of doing common tasks like changing a column's settings; and even with the parts where you'd think "this is so simple, there's no room for variety" such as "select * from table limit 10" - it turns out that "limit" is actually an unofficial extension, not part of the standard, and those databases who support it disagree on how it should be used 😭
